Added support for Sequences (aka macros)

riskable opened the pull request:
Demo: https://youtu.be/M2OXrNP3fgI

Example usage:

```rust
use keyberon::action::{k, l, m, kp, kr, tap, Action, Action::*, SequenceEvent};
use keyberon::key_code::KeyCode::{self, *};

const MACROTEST: Action = Sequence {
    delay: 10,
    actions: &[
        SequenceEvent::Press(LCtrl), SequenceEvent::Press(LShift), SequenceEvent::Press(U), // Long form
        tap(Kb1), tap(F), tap(Kb4), tap(A), tap(F), // Short form
        kr(LCtrl), kr(LShift), kr(U), // Still need to get ReleaseAll() working
    ],
};
```

I couldn't get working `SequenceEvent::ReleaseAll()` (see: https://github.com/riskable/keyberon/blob/master/src/layout.rs#L324) working.  Probably a trivial task for you =)

Also, I'm not sure what the purpose is of that `transform()` method (it's not used anywhere) so I just guessed in my own implementation, haha.  I also didn't implement `Action.key_codes()` for `Action::Sequence` because I'm not sure it's necessary.  I have a little commented out line in place where it'd go.

Overall it wasn't too bad and I learned a ton.  Thanks for the assistance and let me know what I have to change or if it's all bad or whatever =)

TeXitoi commented:
Please fmt your code.

TeXitoi commented on Cargo.toml:
```diff
@@ -3,7 +3,7 @@ name = "keyberon"
 authors = ["Guillaume Pinot <texitoi@texitoi.eu>", "Robin Krahl <robin.krahl@ireas.org>"]
 edition = "2018"
-description = "Parse command line argument by defining a struct."
+description = "Pure Rust keyboard firmware."
```
These are already merged, please rebase

riskable replied:
Doh!  Easy fix.

TeXitoi commented on src/action.rs:
```diff
@@ -1,7 +1,21 @@
+    /// Combination quick keydown followed by keyrelease
+    Tap(KeyCode),
+    /// Release all (currently) pressed keys
+    ReleaseAll(),
```
Let's keep it simple first, and only support Press and Release

riskable replied:
Removed.  That was easy =)

TeXitoi commented on src/action.rs:
```diff
@@ -78,8 +100,33 @@ pub const fn d(layer: usize) -> Action {
+/// A shortcut to create `KeyEvent::Tap`, useful to create compact
+/// layout.
+pub const fn tap(kc: KeyCode) -> SequenceEvent {
+    SequenceEvent::Tap(kc)
```
These shortcut are of little help compared to `use SequenceEvent::*`, and will not be so used. And with the macro we'll add, it will not really be interesting. Let's remove them for now and see at usage that helpers are common. (removing a function is a breaking change, thus we prefer to not put a function if we're not sure it's a good idea).

riskable replied:
This makes sense.  I'll remove them.

TeXitoi commented on src/layout.rs:
```diff
@@ -32,13 +32,18 @@ pub enum Event {
+    /// Press event with just a keycode (for sequences)
+    SequencePress(KeyCode),
+    /// Release event with just a keycode (for sequences)
+    SequenceRelease(KeyCode),
```
You can't modify this enum. It's the public contract of events entering a layout. This enum is not `non_exhaustive`, and thus adding variants is a breaking change.

gilescope replied:
are breaking changes bad for keyboard firmware - I'm build and flash it all in one go so from my side I wouldn't be affected by a breaking change?  - maybe we should break now, up the semver and make it non exhaustive at the same time?

TeXitoi replied:
Breaking change will say that you don't to just do cargo update to update keyberon. I don't care of making breaking changes, but it must add something to the user.

But here, the problem is more that implementation details leak in the public interface.

riskable replied:
@gilescope I think TeXitoi was just making a point (in a nice, technical way) that how I implemented this particular feature sucks, haha.  He's right though!  I've got a much better solution that I'm currently working on that should allow us to incorporate specific delays between keystrokes instead of just one universal delay that's the same for all `Press()` and `Release()` events (in a given macro).

I'm still new to Rust and his wisdom is appreciated :smile: .  I'm currently struggling with writing code (a `map()`, `filter()` or `filiter_map()`) that figures out which keycodes need to remain down during the execution of any given sequence.  It's not so simple because if you have a `Press()` but no corresponding `Release()` until the end of the sequence you need to A) figure out where in the sequence you are and B) cancel out matching pairs of `Press()` and `Release()` events *but only if they're within the same window of the sequence* (that you're currently at).

I actually got the code working to execute simple sequences but it currently doesn't work with modifiers or anything that needs to remain held during the sequence.  Hopefully not much longer now!

Aside: In the process of writing this I've learned a lot about how HID keyboards work, haha.  Like how you can only have a maximum of six keys down at a time--even though there's no real technical reason behind this limitation.  It's just that when folks were defining the USB HID spec they didn't say how many keycodes could be sent at once so *vendors* (e.g. Apple and Microsoft) limited it to six simultaneous keycodes in their implementations (or at least, that's how it was explained to me).  No idea what happens if you try to send more but it's in my TODO list of things to play with =)

TeXitoi replied:
The "usb keyboard boot protocol" has this 6 keys (+8 modifiers) limitation. There is alternative USB descriptor, but the one used in keyberon has this limitation (that I didn't find limiting, so I didn't try the "NKRO" descriptor support).

TeXitoi commented on src/action.rs:
```diff
@@ -41,6 +55,13 @@ pub enum Action {
+        /// How long to delay between events
+        delay: u16, // NOTE: Currently unused
+        /// The sequence of KeyEvents that will be triggered
+        actions: &'static [SequenceEvent],
```
That's not actions, thats keys. `actions` would be if you had `&[Action]`.

riskable replied:
Good catch.  I've renamed it to just `events` in my working code.

TeXitoi commented on src/layout.rs:
```diff
@@ -71,11 +83,13 @@ impl Event {
 enum State {
     NormalKey { keycode: KeyCode, coord: (u8, u8) },
     LayerModifier { value: usize, coord: (u8, u8) },
+    SequenceKey { keycode: KeyCode },
```
The idea is to manage the macro here. for example, we can have a 

```rust
Sequence {
    since: u32, // keep track of timing, incremented at each tick
    delay: u16,
    sequence: &'static [SequenceEvent],
}
```

`since / delay` gives you the index un sequence you're at. Then you can compute the KeyCodes that support this.

if `since / delay > sequence.len()`, tick() would return None, and the state will disapear, and all keys will be automatically released.

riskable replied:
We don't want all keys to be automatically released though.  We want them to only be released when a specific `Release()` is called.

riskable replied:
Just to provide an update (and remind myself where I'm at later because I have to step away now for most of the day)...  I've got an implementation using `SequenceKey` inside of `enum State` but it's only working for simple events where a key doesn't need to be held before another key is pressed.   So it can type out a word like, 'foo' but not 'FOO' (because shift would need to be held the whole time).  It seems that something is generating a release event immediately after every press event.  I might have to setup OpenOCD again and whip out gdb because I'm not not finding it.

I'll keep working on it later today.

TeXitoi replied:
Unit tests are your friends.

TeXitoi commented on src/layout.rs:
```diff
@@ -278,6 +308,36 @@ impl Layout {
+                for key_event in actions {
+                    match *key_event {
+                        SequenceEvent::Press(keycode) => {
+                            self.stacked.push_back(Event::SequencePress(keycode).into());
```
why `.into()`?

riskable replied:
The `into()` is there because it would not compile without it.  It throws this error:

```
   Compiling keyberon v0.1.1 (/home/riskable/workspace/rust/embedded/keyberon)
error[E0308]: mismatched types
   --> /home/riskable/workspace/rust/embedded/keyberon/src/layout.rs:315:52
    |
315 | ...                   self.stacked.push_back(Event::SequencePress(keycode));
    |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `layout::Stacked`, found enum `layout::Event`
```

riskable commented:
All fixed I hope!  I followed your instructions as best I understood them and also made `Delay()` work!  Though to use it you have to do something like this:

```rust
const MACROTEST2: Action = Sequence {
    events: &[
        Press(A), Release(A),
        Delay {since: 0, ticks: 5000},
        Press(F), Release(F),
    ],
};
```
Maybe it could use a shortcut function?

riskable commented:
Just want to mention #10 so this PR gets tracked in that issue.

TeXitoi commented:
Please run fmt and fix clippy warnings.

riskable commented:
I made the changes necessary to make Clippy happy but they weren't to my code...  That's why I left them alone previously (because I didn't want to mess up your stuff--in case you intentionally wanted the code to be as it was).

Anyways, all set!

riskable commented:
I have decided to become the hero we all deserve!  Just pushed something awesome: Not only do we now support sequences but _they can be of nearly limitless length_!  Muwahahaha!

Clippy is still complaining about a "never loop" and I tried a whole bunch of ways to fix it and none of them worked.  It's annoying the heck out of me right now.
![Screenshot_20201019_223415](https://user-images.githubusercontent.com/66987/96533702-565a9400-125c-11eb-986c-d96de040d52f.png)

I've been pestering various Rust chats asking for help but I'm getting a surprising lack of answers.

Tried this but the unit tests fail:
```rust
let chunk_length = self.sequenced.capacity() - 1; // -1: Keep a slot for Continue()
                for key_event in events.chunks(chunk_length).next() {
                    match *key_event {
                        [SequenceEvent::Press(keycode)] => {
                            self.sequenced.push_back(SequenceEvent::Press(keycode));
                        }
                        [SequenceEvent::Release(keycode)] => {
                            self.sequenced.push_back(SequenceEvent::Release(keycode));
                        }
                        [SequenceEvent::Delay { since, ticks }] => {
                            self.sequenced
                                .push_back(SequenceEvent::Delay { since, ticks });
                        }
                        _ => {} // Should never reach a continue here
                    }
                }
                // Add a continuation
                self.sequenced.push_back(SequenceEvent::Continue {
                    index: 0,
                    events,
                });
```

TeXitoi commented:
You still have a lot of not fmted code.

For the clippy warning,
```rust
for e in iter {
    // something
    break;
}
```
is equivalent to
```rust
if let Some(e) = iter.next() {
    // something
}
```

riskable commented:
Those two are not equivalents. I get the same error as I get when I changed it to `for key_event in events.chunks(chunk_length).next()`:
![Screenshot_20201020_084319](https://user-images.githubusercontent.com/66987/96587570-7ae05b80-12b0-11eb-9151-e3fff4853855.png)
I'm still trying to figure it out.

Also, I *did* run rustfmt on the code.  What do you see that isn't formatted?

riskable commented:
Aha!  Thanks to some help in Discord I've successfully made Clippy happy.  I was *so* confused on that one, haha.

I also re-ran both the `rust-analyzer` fmt and `rustfmt` on the code and as far as I can tell the only thing it changed is the `rustfmt` wants a space after elements in the unit test arrays (e.g. `[LCtrl, U]` instead of `[LCtrl,U]` (which `rust-analyzer`'s fmt function doesn't care about).

riskable commented:
In my testing I accidentally added a sequence with a ridiculously long delay (two extra 0s haha) and it occurred to me that there's no mechanism to cancel a running sequence without hitting the reset switch.  To solve that problem I just added a `CancelSequence` action (which can be bound to a key like anything else) and `SequenceEvent::Complete` which can be used instead of several `Release()` actions.  `Complete` is basically the same thing as the `ReleaseAll()` concept we discussed previously except it will work on *any* type of SequenceEvent in case we support stuff like emulated mouse/joystick events in the future.

riskable commented:
Anything else you need me to change?  The code is passing all tests and has had rustfmt run.  I'll change whatever else you want just need some direction :+1:

TeXitoi commented:
Sorry, not so much time to pause and carefully read this PR. Hopefully soon, I will not forget.

riskable commented:
I know you said, "Let's keep it simple first, and only support Press and Release" but I added `SequenceEvent::Tap` just now because I was entering a large-ish sequence and it was annoying me to have to type out `Press()` and `Release()` over and over again for what could be a simple `Tap()`.

Adding support for `Tap()` only required six lines so hopefully that passes the, "let's keep it simple" requirement :smile:

riskable commented:
A (very) gentle reminder to review this PR (really, there's no rush just excitement!).

BTW: I'm working on another PR for Keyberon that adds `TryFrom<u8>` and `TryFrom<char>` support (and unit tests!) to `key_codes::KeyCode`.  It's (my idea of) the first step in making it possible to convert a long string like, "The quick brown fox jumped over the lazy dog." to a `Sequence` of `Tap()` events.  I've got all the alphas/numbers handled via a macro (instead of an enormous `match` block) and I'm slowly working my way through the other edge cases by calculating the enum distance from ASCII's codes (http://www.asciitable.com/).

TeXitoi commented:
You don't want to do these transformations at runtime. It should be more for a crate of macro/procmacro with useful helpers.

TeXitoi commented:
It will also not work for non american keyboard users as is.

TeXitoi commented:
What is "update configuration on the fly"? If you think of something like QMK via, that's basically writing something to flash, so not really "at runtime".

KeyCode are just code that is interpreted by the OS. The official naming of the HID KeyCodes are based on the american layout. That's why the name of the key codes in keyberon reflect the american layout. You have some strangety as `NonUsBackslash`, and that's also the official name of the keycode. If we want to provide easy names corresponding to some other layout, we can have some module with const declaration of the keycode using the corresponding layout. It would be more modular to declare them in a dedicated crate for each layout, limiting the size of keyberon, easily adding layout without modifying keyberon and limiting features in keyberon.

riskable commented:
Let's say I wanted to make a temporary macro that types `yes\n` 5 times with a 2-second delay between repeats.  I only need this for like a day (just doing some installs) so I don't want to update my whole firmware for this.  I send a string like `{"seq": "yes\\n", "repeat": 5, "name": "yes", "row": 1, "col": 16}` over the keyboard's USB serial port.  If there's no way to convert a string (or `char` or `u8`) to `KeyCode`s how can I pull that off?

TeXitoi commented:
For such a case, maybe that's even better to have some kind of recording directly in the keyboard.

As you described, that's a very complicated feature, needing json parsing and layout modification on the fly.

riskable commented:
> For such a case, maybe that's even better to have some kind of recording directly in the keyboard.

Shall I get started on that then? :smile:

TeXitoi commented on src/action.rs:
```diff
@@ -1,7 +1,48 @@
+
+impl SequenceEvent {
+    /// Returns the keycode associated with the given Press/Release event
+    pub fn keycode(&self) -> Option<KeyCode> {
```
That's never used?

riskable replied:
Well, I used it in *debugging* heh heh.  I can get rid of it if you want.  It's not important.  I'm used to Python where metaprogramming and introspection are easy so in Rust I seem to have developed this habit of adding methods of introspection to everything so I can get a view into what's going on whenever I need it.

TeXitoi replied:
derive(Debug) is your friend

TeXitoi commented on src/layout.rs:
```diff
@@ -67,26 +68,49 @@ impl Event {
 enum State {
-    NormalKey { keycode: KeyCode, coord: (u8, u8) },
-    LayerModifier { value: usize, coord: (u8, u8) },
+    /// Normal keys
```
That's an internal structure, not exposed to the world, so no need to doccomment (and these doc comments don't really help)

riskable replied:
Well, you said you wanted to document everything :smile:

I actually added them because Rust was complaining that not everything had a docstring...  So I went to town adding them to everything just to keep it happy.  Though some of them seem important/make sense *to me*.  For example:

```rust
    /// Fake key event for sequences
    FakeKey {
        keycode: KeyCode,
    },
```

I wouldn't want to remove that one because the point of a `FakeKey` isn't obvious.  If I were looking at this code for the first time I'd think, "Why TF is that there?" and it wouldn't immediately occur to me that it's something that's necessary for sequences/macros.

Should it just be a comment?  I'll remove the others.

TeXitoi replied:
You can of course put interesting comments. In this case, better something like "A key generated by some event. Mainly used by SeqenceEvent".

TeXitoi commented on src/layout.rs:
```diff
@@ -185,8 +207,100 @@ impl Layout {
     }
+    /// Takes care of draining and populating the `sequenced` ArrayDeque,
+    /// giving us sequences (aka macros) of nearly limitless length!
+    fn process_sequences(&mut self) {
```
You can really simplify this. You only need one variant at a time, having a deque of 16 is really overkill for memory, and for complexity of code.

You can do everything with

```
    cur_event: Option<Event>,
   awaiting_events: &'static [Event],
```

And when you need another event, something like

```rust
match self.awaiting_events {
    [e, tail @ ..] => {
        self.cur_event = Some(e);
        self.awaiting_event = tail;
    }
    [] => (),
}
```

riskable replied:
I'll see if I can make that work...  I actually tried something like this initially but couldn't get past the borrow checker complaining about not knowing the length of the events ahead of time...  If I only had one sequence I could get it to work but if there were multiple sequences of different lengths it would complain because the size of the events array would be different every time.

TeXitoi replied:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=25221a8bfda08ec6eb055ea5a3efcdb7

You can also have a `ArrayDeque<&'static [Event]>` if you need several slots.

TeXitoi replied:
https://doc.rust-lang.org/stable/std/primitive.slice.html#method.split_first is more appropriate.

TeXitoi commented on src/layout.rs:
```diff
@@ -278,6 +392,59 @@ impl Layout {
+                // Copy the first chunk of `SequenceEvent`s into the `sequenced` ArrayDeque
+                // ...we'll cover the remaining chunks as we drain them from `sequenced`
+                let chunk_length = self.sequenced.capacity() - 1; // -1: Keep a slot for Continue()
+                if let Some(chunk) = events.chunks(chunk_length).next() {
```
My proposition will simplify this.

And what if a sequence is executed during the execution of another sequence? It need a test.

riskable replied:
Yeah, multiple simultaneous sequences is an edge case in the TODO list.  The way it works right now is that if you kick off a sequence while another one is already running it ends the current sequence and starts the new one.  I made the `sequenced` `ArrayDeque` 16 items long originally so I could have multiple simultaneous sequences running but the way the logic works right now that can't work.  I had planned to use the length of `sequenced` divided by four so you could have up to four simultaneous sequences going on but the amount of logic that would need to change for that to happen was too much (for the current state of the PR--can't have everything at the start of these things).

TeXitoi replied:
You can document the actual behavior, and add an issue.

riskable commented on src/action.rs:
```diff
@@ -1,7 +1,35 @@
+        /// A delay (in ticks) to wait before executing the next SequenceEvent
+        since: u32,
+        /// Number of ticks to wait before removing the Delay
+        ticks: u32,
```
`since` gets updated with each regular call to `Layout.tick()`; it's the timekeeper (like everything else that uses `since`).  `ticks` tells us how long the `SequenceEvent::Delay` should last.  As soon as `since >= ticks` we know the `Delay` is done and can allow it to fall out of the `sequenced` queue.

The two values keep getting pulled and pushed back into `sequenced` until the `Delay` is done which is why they go together like that.

TeXitoi replied:
Just have a `duration: u16` field, and decrement it until 0.

riskable commented on src/layout.rs:
```diff
@@ -185,6 +214,32 @@ impl Layout {
+                        // Increment and put it back
+                        self.sequenced.push_front(SequenceEvent::Delay {
+                            since: since.saturating_add(1),
+                            ticks: 1, // Count this tick as the first
```
This is outdated.  I fixed this bug already in ab0bfe1

riskable commented on src/action.rs:
```diff
@@ -1,7 +1,43 @@
+    },
+    /// A marker that indicates there's more of the macro than would fit
+    /// in the 'sequenced' ArrayDeque
+    Continue {
```
I'm not sure I can put it anywhere else considering that the `sequence` `ArrayDeque` only takes the `SequenceEvent` type.  By having it here the logic of draining and topping off `sequenced` stays nice and simple.

If I were to put it somewhere else I'd need to add a parallel data structure that mirrors `sequenced` and then doubly keep track of both `sequenced` and that new ArrayDeque (or some sort of array that tracks sequence markers or similar) to make sure the `Continue` matches everything 1-for-1.

When I first started writing this I actually had something similar that kept track of the indexes/positions/delays of everything in separate objects but the logic grew so complicated I couldn't keep it all in my working memory anymore so I abandoned it in favor of this method.

TeXitoi replied:
That's the enum exposed to the user, so it must not contain any implementation detail.

TeXitoi replied:
Having a clean public interface allow easier configuration for the users, and internal refactoring not breaking the interface.

TeXitoi replied:
This is not fixed

TeXitoi commented:
Any problem with the revîew? Do you need some help to fix it?

I hope I didn't discourage you with this review :innocent:

riskable commented:
You haven't discouraged me... It just took a while for the initial review that I started work on other things :smile: 

I'm finishing up those other things (tease: https://imgur.com/a/IFsvBGh) and will be returning to this in the next few days.  I *did* start work on it though... Already implemented a decrementing counter instead of using the since/ticks method :+1:

riskable commented:
I'm done making my keyboard PCB and was able to get back to work on this!  I implemented all of your suggested changes and added some general improvements (can handle up to four simultaneous running sequences now!).  Had a hell of a time with the merge from upstream though haha.  Hopefully I didn't clobber anything! Double-check that please :smile: 

I've got complete unit tests for everything except multiple simultaneous sequences running at the same time because I'm not sure how to handle that (in test cases).

riskable commented:
Just checking in...  In case you missed Github's notification that I've submitted a new PR :grin:

riskable commented:
I just merged upstream (everything's up to date again).  Any chance you can take a look?

TeXitoi commented:
I really want a clean public interface to merge it. Implementation details can be changed later, so OK to merge without the other proposed change.
